### PR TITLE
Update relative position to be based on lockable and has_lesson_plan

### PR DESCRIPTION
**Pull request by dmcavoy**
## Background

Previous PRs have:
- Added has_lesson_plan to lesson model
- Set has_lesson_plan for existing lessons based on lesson lockable and script has_lesson_plan setting

## This PR

Updates relative position to be determined based on lockable and has_lesson plan.  Relative position is the position relative to other things in the same category. The categories are 'lesson' and 'survey'.  Survey is a new word for something that is lockable without a lesson plan.  After talking to the curriculum team most of the time something that is lockable without a lesson plan is a survey. See the table below explaining the difference. 

| Lockable ? | Has Lesson Plan ? | Survey or Lesson ? |
| -- | -- | -- |
| True | True | Lesson |
| True | False | Survey |
| False | True | Lesson |
| False | False | Lesson

The script_level url will be determined based on these new survey and lesson categories. Surveys will go to /lockable/xx/puzzle/xx and lessons with go to /stage/xx/puzzle/xx. 

The only lessons which are impacted by the relative position change currently are in CSP 21-22 which is not yet being used. A future PR will deal with updating CSP 20-21 lessons that are lockable and have lesson plans. 

## Future Work

- Update the display of lesson names to show 'Lesson #' for lessons that are lockable but have a lesson plan (https://github.com/code-dot-org/code-dot-org/pull/38746)
- Deal with the CSP 20-21 lessons that are lockable and have lesson plans
- Updating to new URLs (/survey and /lesson)

## Links

- [spec](https://docs.google.com/document/d/1sAAVnwv3rXQEK-qa8JnkNXFfZ0vjWQyzJ9Yx_pG4bdg/edit#)
- [jira](https://codedotorg.atlassian.net/browse/PLAT-722)

## Testing story

I manually tested some different cases locally.

The two tests that I have updated that seem most important to showing that this is working:
- levels_helper_test.rb
- lesson_test.rb

Beyond that I'm counting on existing tests.

### CODE REVIEW

File: aws/redshift/views/course_structure.sql [hunk 1]
```diff
@@ -12,7 +12,7 @@ select
   sc.name script_name, 
   st.id stage_id, 
   st.name stage_name, 
-  case when lockable = 1 then st.absolute_position else st.relative_position end stage_number, 
+  case when lockable = 1 AND has_lesson_plan = 0 then st.absolute_position else st.relative_position end stage_number,
```
Comment (dmcavoy) [inline aws/redshift/views/course_structure.sql:15]: @bakerfranke Just tagging you here to double check that this is alright to change.

File: dashboard/app/models/lesson.rb [hunk 2]
```diff
@@ -94,7 +95,7 @@ def self.add_lessons(script, lesson_group, raw_lessons, counters, new_suffix, ed
         lockable: !!raw_lesson[:lockable],
         has_lesson_plan: !!raw_lesson[:has_lesson_plan],
         visible_after: raw_lesson[:visible_after],
-        relative_position: !!raw_lesson[:lockable] ? (counters.lockable_count += 1) : (counters.non_lockable_count += 1)
+        relative_position: !!raw_lesson[:has_lesson_plan] || !raw_lesson[:lockable] ? (counters.lesson_count += 1) : (counters.survey_count += 1)
```
Comment (bethanyaconnor) [inline dashboard/app/models/lesson.rb]: nit: would it be easier to reverse this ternary? I personally feel like it's easier to check for the simplest case (what lessons are surveys) than the three other cases. But happy to be overruled here

File: dashboard/app/models/script.rb [hunk 3]
```diff
@@ -823,15 +827,15 @@ def get_script_level_by_id(script_level_id)
...
+  def get_script_level_by_relative_position_and_puzzle_position(relative_position, puzzle_position, survey)
     relative_position ||= 1
     script_levels.find do |sl|
       # make sure we are checking the native properties of the script level
       # first, so we only have to load lesson if it's actually necessary.
       sl.position == puzzle_position.to_i &&
         !sl.bonus &&
         sl.lesson.relative_position == relative_position.to_i &&
-        sl.lesson.lockable? == lockable
+        (survey ? sl.lesson.lockable? == true && sl.lesson.has_lesson_plan? == false : sl.lesson.lockable? == false || sl.lesson.has_lesson_plan? == true)
```
Comment (bethanyaconnor) [inline dashboard/app/models/script.rb]: nit: the `== true/false` parts of the clause aren't helpful to me and clutter this statement. This definitely could be personal preference though so if they make it clearer to others, feel free to ignore

File: dashboard/config/scripts_json/csp9-2021.script_json [hunk 4]
```diff
@@ -274,7 +274,7 @@
       "absolute_position": 9,
       "lockable": true,
       "has_lesson_plan": true,
-      "relative_position": 1,
+      "relative_position": 9,
```
Comment (bethanyaconnor) [inline dashboard/config/scripts_json/csp9-2021.script_json:277]: out of curiousity, how did you generate the list of lessons to modify the positions of?

**Review by bethanyaconnor — APPROVED** [review]
A couple of inline nits but this looks good to me. I'm not familiar with any potential traps here though

File: dashboard/app/models/script.rb [hunk 5]
```diff
@@ -336,10 +336,14 @@ def starting_level
...
   end
 
-  # Find the lockable or non-locakble stage based on its relative position.
-  # Raises `ActiveRecord::RecordNotFound` if no matching stage is found.
-  def stage_by_relative_position(position, lockable = false)
-    lessons.where(lockable: lockable).find_by!(relative_position: position)
+  # Find the lesson based on its relative position, lockable value, and if it has a lesson plan.
+  # Raises `ActiveRecord::RecordNotFound` if no matching lesson is found.
+  def lesson_by_relative_position(position, survey = false)
+    if survey
```
Comment (davidsbailey) [inline dashboard/app/models/script.rb]: would you be willing to call this `use_survey_url` instead? I still think `/survey/` is a good choice for the url, but I am worried that it is a bit confusing to use this term like this in the code without explaining it, given that "survey" already has a definition in our system, i.e. an anonymous level group.

File: dashboard/app/models/lesson.rb (hunk 2 above)
Comment (davidsbailey) [reply dashboard/app/models/lesson.rb]: this does seem a bit hard to follow. I would suggest something like:
```
use_survey_url = ...
```
and then
```
relative_position: use_survey_url ? ... : ...
```
alternately, `use_survey_url` could be extracted as a method.

File: dashboard/app/models/script.rb (hunk 3 above)
Comment (davidsbailey) [reply dashboard/app/models/script.rb]: could some of this new logic be extracted into a variable with a descriptive name? I think this would be very hard for someone to understand without a lot of context.

**Review by davidsbailey — COMMENTED** [review]
some initial comments

File: dashboard/config/scripts_json/csp9-2021.script_json (hunk 4 above)
Comment (dmcavoy) [reply dashboard/config/scripts_json/csp9-2021.script_json:277]: When I set has_lesson_plan for everything it was set based on lockable.  So if lockable was true then has_lesson_plan was false. I then updated the lessons in CSP 21-22 that we know are lockable but have lesson plans to set has_lesson_plan: true. So these are the lessons that would be impacted by this update to how we decided which URL because any other condition stays with the same url as before.  There are lessons in CSP 20-21 that also need this special update but because they are live in use we need to do a little more work to make that change.

File: dashboard/app/models/script.rb [hunk 6]
```diff
@@ -823,15 +827,15 @@ def get_script_level_by_id(script_level_id)
...
+  def get_script_level_by_relative_position_and_puzzle_position(relative_position, puzzle_position, use_survey_url)
     relative_position ||= 1
     script_levels.find do |sl|
       # make sure we are checking the native properties of the script level
       # first, so we only have to load lesson if it's actually necessary.
       sl.position == puzzle_position.to_i &&
         !sl.bonus &&
         sl.lesson.relative_position == relative_position.to_i &&
-        sl.lesson.lockable? == lockable
+        (use_survey_url ? !sl.lesson.numbered_lesson? : sl.lesson.numbered_lesson?)
```
Comment (davidsbailey) [inline dashboard/app/models/script.rb]: thank you, this is much better! however, I think there is still room for improvement:
1. this looks like it can be further simplified to `use_survey_url == !sl.lesson.numbered_lesson?`
2. even after addressing (1), it is still a bit more complicated than it needs to be to have two concepts which are opposites, roughly "survey" and "numbered". would you be willing to pick one of these and stick to it, so that you can go back to having something which looks more like the original code? then this line could read: `is_numbered_lesson == sl.lesson.numbered_lesson?` or `use_survey_url == sl.lesson.use_survey_url?`
3. it would be helpful to standardize on "survey" or "numbered" throughout the code you are touching in this PR.

File: dashboard/config/scripts_json/csp9-2021.script_json (hunk 4 above)
Comment (bethanyaconnor) [reply dashboard/config/scripts_json/csp9-2021.script_json:277]: Thanks!

File: dashboard/app/models/script.rb (hunk 6 above)
Comment (dmcavoy) [reply dashboard/app/models/script.rb]: Thanks Dave for the push to make this change. Its definitely the right thing to update this. I have taken a pass out this. I used "unnumbered" and "numbered" lesson but let me know if you think there is a better wording to use

File: dashboard/app/models/lesson.rb [hunk 7]
```diff
@@ -63,8 +63,10 @@ class Lesson < ApplicationRecord
   )
 
   # A lesson has an absolute position and a relative position. The difference between the two is that relative_position
-  # only accounts for other lessons that have the same lockable setting, so if we have two lockable lessons followed
-  # by a non-lockable lesson, the third lesson will have an absolute_position of 3 but a relative_position of 1
+  # numbers the lessons in order in two groups 1. lessons that are numbered (lockable false OR has_lesson_plan true)
+  # 2. lessons that are not numbered (lockable true AND has_lesson_plan false)
+  # if we have two lessons without lesson plans that are lockable followed by a
+  # lesson that is not lockable, the third lesson will have an absolute_position of 3 but a relative_position of 1
```
Comment (davidsbailey) [inline dashboard/app/models/lesson.rb:69]: numbered and unnumbered sounds great! I would just add in the comments here that "numbered" means that the display name is prefixed with the lesson number, or some other wording -- otherwise it was a bit confusing because I thought it was referring to urls at first, where "unnumbered" lessons are still numbered.

File: dashboard/app/models/lesson.rb [hunk 8]
```diff
@@ -171,6 +175,11 @@ def spelling_bee?
     script_levels.first.oldest_active_level.spelling_bee?
   end
 
+  # We number lessons that either have lesson plans or are not lockable
+  def numbered_lesson?
+    has_lesson_plan || !lockable
```
Comment (davidsbailey) [inline dashboard/app/models/lesson.rb]: I guess if we are going to be comparing the output of this using == to another boolean, then this should start with `!!has_lesson_plan`

**Review by davidsbailey — APPROVED** [review]
a couple more nits, but otherwise this is looking good! Thank you for persevering on this, Dani!
